### DISCUSSION TITLE
How can I add extra params to the LSP client during the runtime?

### DISCUSSION THREAD

**Discussion by 0xDmtri** [VonHeikemen/lsp-zero.nvim#252] [Q&A] [answered]
Hi,

I am developing a small plugin for Solidity and am wondering how can I inject some settings to the runtime without overwriting everything else? I tried the following way:

<img width="522" alt="Screenshot 2023-05-15 at 6 57 58 PM" src="https://github.com/VonHeikemen/lsp-zero.nvim/assets/33504992/4197ea9e-d0c8-46cb-aef9-43b4d802da30">

But what happens is that everything is being overwritten to default except of the field that I have passed thru. Any chance I can exclusively add some settings and keep the rest as per my setup of lsp-zero? Many thanks!

**Comment by VonHeikemen** [comment]
Do you mean after the LSP has started?

**Reply by 0xDmtri** [reply]
yeap

**Reply by VonHeikemen** [reply, marked as answer]
I believe you'll have to send a "notification" to the LSP server. You'll need to do something like this.

```lua
local client = vim.lsp.get_active_clients({name = 'solidity'})[1]

if client == nil then
  return
end

local defaults = client.config.settings

local lines = lines_from(file_name)
local remap_tab = get_tab(lines)

local new_settings = {
  solidity = {
    includePath = '',
    remapping = remap_tab
  }
}

local ok, err = pcall(client.notify, 'workspace/didChangeConfiguration', {
  settings = vim.tbl_deep_extend('force', defaults, new_settings) 
})

if not ok then
  print('Could not send settings')
end
```

For this to work your LSP server has to implement `workspace/didChangeConfiguration`.

**Reply by 0xDmtri** [reply]
thank you! will try now!

**Reply by 0xDmtri** [reply]
> I believe you'll have to send a "notification" to the LSP server. You'll need to do something like this.
> 
> ```lua
> local client = vim.lsp.get_active_clients({name = 'solidity'})[1]
> 
> if client == nil then
>   return
> end
> 
> local defaults = client.config.settings
> 
> local lines = lines_from(file_name)
> local remap_tab = get_tab(lines)
> 
> local new_settings = {
>   solidity = {
>     includePath = '',
>     remapping = remap_tab
>   }
> }
> 
> local ok, err = pcall(client.notify, 'workspace/didChangeConfiguration', {
>   settings = vim.tbl_deep_extend('force', defaults, new_settings) 
> })
> 
> if not ok then
>   print('Could not send settings')
> end
> ```
> 
> For this to work your LSP server has to implement `workspace/didChangeConfiguration`.

Unfortunately Solidity does not have `workspace/didChangeConfiguration`. I am now thinking what if I follow similar logic but instead of `local defaults = client.config.settings` I save the entire thing like `local config = client.config` and then, add my part that is generated with my code via `config.settings = new_settings`, so after I can call `lsp_zero.configure('solidity' , config)`?

**Reply by VonHeikemen** [reply]
Next best thing would be to setup your plugin before calling the solidity LSP with lspconfig.

You would have this piece of code in a function.

```lua
local lsp_util = require('lspconfig.util')

lsp_util.on_setup = lsp_util.add_hook_after(
  lsp_util.on_setup,
  function(config)
    if config.name ~= 'solidity' then
      return
    end

    local lines = lines_from(file_name)
    local remap_tab = get_tab(lines)

    local new_settings = {
      solidity = {
        includePath = '',
        remapping = remap_tab
      }
    }

    config.settings = vim.tbl_deep_extend(
      'force',
      config.settings,
      new_settings
    )
  end
)
```

The users of your plugin would have to execute your setup before calling the server.

```lua
my_plugin_setup()

require('lspconfig').solidity.setup({})
```
